csrc/kernel_cache.h

```diff
@@ -213,6 +213,24 @@ class TORCH_CUDA_CU_API FusionKernelRuntime {
     return executors_;
   }
 
+  std::vector<at::Tensor> allocOutputSpace(KernelArgumentHolder& args) {
+    if (!isCompiled()) {
+      compileFusionParallel(args);
+    }
+
+    std::vector<at::Tensor> outputs;
+    for (auto& executor : executors_) {
```

Comment:
Is it guaranteed that if we run executors sequentially, we will get the outputs in the expected order? I don't think this is correct, for example, if we have two segments vertically, the the first executor's output will be fully consumed by the second executor, and shouldn't be used as the output of the entire fusion. I think it makes more sense to pass in a vector of `TensorView*`s specifying what output you want to allocate, then only allocate that output.

Reply:
I'm also not sure about this part. Is the intent to allocate the outputs of the fusion? I think that's what you want to do here, so it doesn't seem to necessary to allocate the outputs for all segments. Note that a fusion is divided into segments, and each segment has its outputs, which are consumed by subsequent segments. From the perspectives of the pipeline execution, it shouldn't need to bother with how segments are executed, so just allocating buffers for the fusion outputs should be enough, right?

Reply:
Oh right ! I really wasn't sure of what I was doing so thanks for spotting this mistake. It is not so easy for me to deal with the execution logic.

> just allocating buffers for the fusion outputs should be enough, right?

Yes indeed ! How can we achieve that?

In the meantime I propose a cheap and easy fix, which is to run the stage (regardless of the result of `shouldRun`). It will indeed allocate the outputs (and fill them with corrupted data, but those corrupted data won't be used)

Reply:
I am OK with this temporary garbage-in-garbage-out approach for this PR. We need to rework this allocation part though. There are two separate questions:
1. Do we do garbage computation?
2. Do we allocate garbage memory?
Even if we correctly implemented the `FusionExecutorCache::allocOutputSpace`, we are only avoiding 1, but we are still doing 2. So I don't mind having 1 at this moment, and in the future, we should resolve both 1 and 2.

Reply:
Yes to 1 and 2. I agree with your comment. There is one "TODO" comment for each 1. and 2. inside `multidevice/executor.cpp`

To avoid 2. we would need to alloc output only if strictly necessary, in which case we still need to now what should be the allocation size. So It would be great to have a function like
`type? FusionExecutorCache::InferOutputShape(type? input_shape)`